rngd.8
.\" Copyright (C) 2001 Jeff Garzik -- jgarzik@pobox.com
.\"
.TH RNGD 8 "March 2001" "rng-tools 6"

.SH NAME
rngd \- Check and feed random data from hardware device to kernel random device

.SH SYNOPSIS
.B rngd
[\fB\-b\fR, \fB\-\-background\fR]
[\fB\-f\fR, \fB\-\-foreground\fR]
[\fB\-d\fR, \fB\-\-debug\fR]
[\fB\-l\fR, \fB\-\-list\fR]
[\fB\-x\fR, \fB\-\-exclude\fR]
[\fB\-i\fR, \fB\-\-ignorefail\fR]
[\fB\-o\fR, \fB\-\-random-device=\fIfile\fR]
[\fB\-p\fR, \fB\-\-pid-file=\fIfile\fR]
[\fB\-r\fR, \fB\-\-rng-device=\fIfile\fR]
[\fB\-s\fR, \fB\-\-random-step=\fInnn\fR]
[\fB\-W\fR, \fB\-\-fill-watermark=\fInnn\fR]
[\fB\-q\fR, \fB\-\-quiet\fR]
[\fB\-?\fR, \fB\-\-help\fR]
[\fB\-V\fR, \fB\-\-version\fR]
.RI

.SH DESCRIPTION
This daemon feeds data from a random number generator to the kernel's
random number entropy pool, after first checking the data to ensure that
it is properly random.
.PP
The \fB\-f\fR or \fB\-\-foreground\fR options can be used to tell
\fBrngd\fR to avoid forking on startup.  This is typically used for
debugging.  The \fB\-b\fR or \fB\-\-background\fR options, which fork and put
\fBrngd\fR into the background automatically, are the default.
.PP
The \fB\-r\fR or \fB\-\-rng-device\fR options can be used to select an
alternate source of input, besides the default /dev/hwrng.
The \fB\-o\fR or \fB\-\-random-device\fR options can be used to select
an alternate entropy output device, besides the default /dev/random.
Note that this device must support the Linux kernel /dev/random 
ioctl API.
.PP

.SH OPTIONS
.TP
\fB\-b\fR, \fB\-\-background\fR
Become a daemon (default)
.TP
\fB\-f\fR, \fB\-\-foreground\fR
Do not fork and become a daemon
.TP
\fB\-d\fR, \fB\-\-debug\fR
Enable debug messages
.TP
\fB\-l\fR, \fB\-\-list\fR
List the indexes and names of available entropy sources. Exit code when listing
sources is 0 if at least 1 entropy source was found and initialized, 1
otherwise.
.TP
\fB\-x\fR, \fB\-\-exclude\fR
Disable entropy sources based on index reported from --list option
.TP
\fB\-p\fI file\fR, \fB\-\-pid-file=\fIfile\fR
File used for recording daemon PID, and multiple exclusion
(default: /var/run/rngd.pid)
.TP
\fB\-i\fR, \fB\-\-ignorefail\fR
Ignore repeated fips failures
.TP
\fB\-o\fI file\fR, \fB\-\-random-device=\fIfile\fR
Kernel device used for random number output
(default: /dev/random)
.TP
\fB\-r\fI file\fR, \fB\-\-rng-device=\fIfile\fR
Kernel device used for random number input
(default: /dev/hwrng)
.TP
\fB\-s\fI nnn\fR, \fB\-\-random-step=\fInnn\fR
Number of bytes written to random-device at a time (default: 64)
.TP
\fB\-e\fI nnn\fR, \fB\-\-entropy-count=\fInnn\fR
Number of bits to consider random when adding entropy. 1<=e<=8
(default: 8)
.TP
\fB\-W\fI n\fR, \fB\-\-fill\-watermark=\fInnn\fR
Once we start doing it, feed entropy to \fIrandom-device\fR until at least
\fIfill-watermark\fR bits of entropy are available in its entropy pool.
By default, this value is set to 75% of the entropy pool size or 2048 bits
if the entropy pool size couldn't be determined.
Setting this too high will cause \fIrngd\fR to dominate the contents of the
entropy pool.  Low values will hurt system performance during entropy 
starves.  Do not set \fIfill-watermark\fR above the size of the
entropy pool (usually 4096 bits).
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Suppress error messages
.TP
\fB\-?\fR, \fB\-\-help\fR
Give a short summary of all program options.
.TP
\fB\-V\fR, \fB\-\-version\fR
Print program version

.SH
ENTROPY SOURCES
.TP
.B
Hardware RNG [Index 0]
The Hardware RNG is a generic entropy source that draws entropy from a hardware
rng attached by default to /dev/hwrng

.TP
.B
TPM [Index 1]
Entropy drawn from on board TPM device

.TP
.B
RDRAND/RDSEED [Index 2]
Entropy drawn from the RDRAND and RDSEED instructions (x86 only)
.TP
Options

\fBuse_aes - \fR select if rdrand is used to seed a cprng to produce entropy, or
if all entropy is drawn directly from rdrand instruction (default 1)

.TP
.B
DARN [Index 3]
Entropy drawn from the DARN instruction (ppc64 only)
.TP
Options

\fBuse_aes - \fR select if the power darn instruction is used to directly fill
the entropy poll, or if it is used as a periodic seed to an aes based cprng
(default 1)

.TP
.B
NIST Entropy Beacon [Index 4]
Entropy gathered from the NIST network entropy beacon.  Note that this entropy
source is disabled by default as it should never be used for cryptographic
purposes, or any use case in which random data should be known only by a single
entity.

.TP
.B
JITTER [Index 5]
Entropy gathered by measuring timing variance of operations on the local cpu
.TP
Options
\fBthread_count - \fR Set the number of threads that the jitter entropy source will spawn

\fBbuffer_size - \fR Set the size of the buffer cache that each thread maintains 

\fBrefill_thresh - \fR Set the low watermark for the cache that triggers its corresponding thread to start refill operations 


.SH AUTHORS
Philipp Rumpf
.br
Jeff Garzik \- jgarzik@pobox.com
.br
Matt Sottek
.br
Brad Hill
.br
Neil Horman - nhorman@tuxdriver.com


NEWS

Version 5 (April 1, 2014):
* Support RDRAND capable systems that don't have AES-NI
* Man page spelling fixes
* Add BUGS document
* Don't use fixed AES key for data reduction
* Platform: Support x32. x86-64 micro-optimizations.
* Fix RDRAND data reduction
* Enable RDSEED instruction

Version 4 (August 2, 2012):
* Add RDRAND instruction support
* Add -q and -v options for quiet and verbose output
* Add -p option for specifying PID file (text file containing daemon's PID)
* Disable entropy source if facing continued failures, but be
  tolerant of the occasional fault.
* Default device is now the preferred /dev/hwrng
* Do not use TPM device for RNG access, if /dev/hwrng is present
  (TPM RNG is exported via the kernel, in newer kernels)

Version 3 (July 3, 2010):

* add rngtest program
* imported into git repository
* support TPM chip's hardware RNG (and thus, a framework for supporting
  multiple entropy sources)
* change default hardware RNG device name to "/dev/hw_random"

Version 2 (August 24, 2004):

* Fixes and cleanups from Debian (Henrique de Moraes Holschuh)
  Assures rngd works with 2.6 kernels.

Version 1.1 (April 6, 2004):

* update to recent autoconf/automake
* add new rngtest program
* various minor cleanups
* much better FIPS testing



rngd.c
/*
 * rngd.c -- Random Number Generator daemon
 *
 * rngd reads data from a hardware random number generator, verifies it
 * looks like random data, and adds it to /dev/random's entropy store.
 *
 * In theory, this should allow you to read very quickly from
 * /dev/random; rngd also adds bytes to the entropy store periodically
 * when it's full, which makes predicting the entropy store's contents
 * harder.
 *
 * Copyright (C) 2001 Philipp Rumpf
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Suite 500, Boston, MA  02110-1335  USA
 */

#define _GNU_SOURCE

#ifndef HAVE_CONFIG_H
#error Invalid or missing autoconf build environment
#endif

#include "rng-tools-config.h"

#include <unistd.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <errno.h>
#include <string.h>
#include <argp.h>
#include <syslog.h>
#include <signal.h>
#include <limits.h>

#include "rngd.h"
#include "fips.h"
#include "exits.h"
#include "rngd_entsource.h"
#include "rngd_linux.h"

/*
 * Globals
 */

/* Background/daemon mode */
bool am_daemon;				/* True if we went daemon */
bool msg_squash = false;		/* True if we want no messages on the console */
bool server_running = true;		/* set to false, to stop daemon */

bool ignorefail = false; /* true if we ignore MAX_RNG_FAILURES */

/* Command line arguments and processing */
const char *argp_program_version =
	"rngd " VERSION "\n"
	"Copyright 2001-2004 Jeff Garzik\n"
	"Copyright 2017 Neil Horman\n"
	"Copyright (c) 2001 by Philipp Rumpf\n"
	"This is free software; see the source for copying conditions.  There is NO "
	"warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.";

const char *argp_program_bug_address = PACKAGE_BUGREPORT;

static char doc[] =
	"Check and feed random data from hardware device to kernel entropy pool.\n";

static struct argp_option options[] = {
	{ "debug", 'd', 0, 0, "Enable debug output" },

	{ "foreground",	'f', 0, 0, "Do not fork and become a daemon" },

	{ "ignorefail", 'i', 0, 0, "Ignore repeated fips failures" },

	{ "background", 'b', 0, 0, "Become a daemon (default)" },

	{ "exclude", 'x', "n", 0, "Disable the numbered entropy source specified" },

	{ "include", 'n', "n", 0, "Enable the numbered entropy source specified" },

	{ "list", 'l', 0, 0, "List the operational entropy sources on this system and exit" },

	{ "option", 'O', "options", 0, "rng specific options in the form source:key:value"},

	{ "random-device", 'o', "file", 0,
	  "Kernel device used for random number output (default: /dev/random)" },

	{ "rng-device", 'r', "file", 0,
	  "Kernel device used for random number input (default: /dev/hwrng)" },

	{ "pid-file", 'p', "file", 0,
	  "File used for recording daemon PID, and multiple exclusion (default: /var/run/rngd.pid)" },

	{ "random-step", 's', "nnn", 0,
	  "Number of bytes written to random-device at a time (default: 64)" },

	{ "fill-watermark", 'W', "n", 0,
	  "Do not stop feeding entropy to random-device until at least n bits of entropy are available in the pool (default: 2048), 0 <= n <= 4096" },

	{ "quiet", 'q', 0, 0, "Suppress error messages" },

	{ "version" ,'v', 0, 0, "List rngd version" },

	{ "entropy-count", 'e', "n", 0, "Number of entropy bits to support (default: 8), 1 <= n <= 8" },

	{ 0 },
};

static struct arguments default_arguments = {
	.random_name	= "/dev/random",
	.pid_file	= "/var/run/rngd.pid",
	.random_step	= 64,
	.daemon		= true,
	.list		= false,
	.ignorefail	= false,
	.quiet		= false,
	.entropy_count	= 8,
};
struct arguments *arguments = &default_arguments;

static enum {
	ENT_HWRNG = 0,
	ENT_TPM = 1,
	ENT_RDRAND,
	ENT_DARN,
	ENT_NISTBEACON,
	ENT_JITTER,
	ENT_MAX
} entropy_indexes;


static struct rng_option drng_options[] = {
	[DRNG_OPT_AES] = {
		.key = "use_aes",
		.int_val = 1,
	},
	{
		.key = NULL,
	},
};

static struct rng_option darn_options[] = {
	[DARN_OPT_AES] = {
		.key = "use_aes",
		.int_val = 1,
	},
	{
		.key = NULL,
	}
};

static struct rng_option jitter_options[] = {
	[JITTER_OPT_THREADS] = {
		.key = "thread_count",
		.int_val = 4,
	},
	[JITTER_OPT_BUF_SZ] = {
		.key = "buffer_size",
		.int_val = 16535,
	},
	[JITTER_OPT_REFILL] = {
		.key = "refill_thresh",
		.int_val = 16535,
	},
};

static struct rng entropy_sources[ENT_MAX] = {
	/* Note, the special char dev must be the first entry */
	{
		.rng_name	= "Hardware RNG Device",
		.rng_fname      = "/dev/hwrng",
		.rng_fd         = -1,
		.xread          = xread,
		.init           = init_entropy_source,
		.rng_options	= NULL,
	},
	/* must be at index 1 */
	{
		.rng_name	= "TPM RNG Device",
		.rng_fname      = "/dev/tpm0",
		.rng_fd         = -1,
		.xread          = xread_tpm,
		.init           = init_tpm_entropy_source,
		.rng_options	= NULL,
	},
	{
		.rng_name       = "Intel RDRAND Instruction RNG",
		.rng_fd         = -1,
#ifdef HAVE_RDRAND
		.xread          = xread_drng,
		.init           = init_drng_entropy_source,
#else
		.disabled	= true,
#endif
		.rng_options	= drng_options,
	},
	{
		.rng_name       = "Power9 DARN Instruction RNG",
		.rng_fd         = -1,
#ifdef HAVE_DARN
		.xread          = xread_darn,
		.init           = init_darn_entropy_source,
#else
		.disabled	= true,
#endif
		.rng_options	= darn_options,
	},
	{
		.rng_name	= "NIST Network Entropy Beacon",
		.rng_fd		= -1,
#ifdef HAVE_NISTBEACON
		.xread		= xread_nist,
		.init		= init_nist_entropy_source,
#endif
		.disabled	= true,
		.rng_options	= NULL,
	},
	{
		.rng_name	= "JITTER Entropy generator",
		.rng_fd		= -1,
#ifdef HAVE_JITTER
		.xread		= xread_jitter,
		.init		= init_jitter_entropy_source,
		.close		= close_jitter_entropy_source,
#else
		.disabled	= true,
#endif
		.rng_options	= jitter_options,
	},
};


/*
 * command line processing
 */
static error_t parse_opt (int key, char *arg, struct argp_state *state)
{
	char *optkey;
	long int idx;
	long int val;
	char *search, *last_search;
	struct rng_option *options;

	switch(key) {
	case 'd':
		arguments->debug = true;
		break;
	case 'o':
		arguments->random_name = arg;
		break;
	case 'O':

		search = strchr(arg, ':');
		if (!search) {
			printf("Invalid rng option format\n");
			return -ERANGE;
		}

		idx = strtoul(arg, &search, 10);
		if ((idx == LONG_MAX) || (idx >= ENT_MAX)) {
			printf("option index out of range: %lu\n", idx);
			return -ERANGE;
		}

		last_search = search + 1;
		search = strchr(last_search, ':');
		if (!search) {
			printf("Available options for %s\n", entropy_sources[idx].rng_name);
			options = entropy_sources[idx].rng_options;
			while (options && options->key) {
				printf("key: [%s]\tdefault value: [%d]\n", options->key, options->int_val);
				options++;
			}
			return -ERANGE;
		}
		*search = '\0';
		optkey = strdupa(last_search);
		*search = ':';

		last_search = search + 1;

		val = strtoul(last_search, NULL, 10);
		if (val == LONG_MAX) {
			printf("rng option was not parsable\n");
			return -ERANGE;
		}

		options = entropy_sources[idx].rng_options;
		while (options && options->key) {
			if (!strcmp(optkey, options->key)) {
				options->int_val = val;
				return 0;
			}
			options++;
		}
		printf("Option %s not found for source idx %d\n", key, idx);
		return -ERANGE;
		break;

	case 'x':
		idx = strtol(arg, NULL, 10);
		if ((idx == LONG_MAX) || (idx >= ENT_MAX)) {
			printf("exclude index is out of range: %lu\n", idx);
			return -ERANGE;
		}
		entropy_sources[idx].disabled = true;
		printf("Disabling %lu: %s\n", idx, entropy_sources[idx].rng_name);
		break;
	case 'n':
		idx = strtol(arg, NULL, 10);
		if ((idx == LONG_MAX) || (idx >= ENT_MAX)) {
			printf("enable index is out of range: %lu\n", idx);
			return -ERANGE;
		}
		entropy_sources[idx].disabled = false;
		printf("Enabling %lu: %s\n", idx, entropy_sources[idx].rng_name);
		break;
	case 'l':
		arguments->list = true;
		break;
	case 'p':
		arguments->pid_file = arg;
		break;
	case 'r':
		entropy_sources[ENT_HWRNG].rng_fname = arg;
		break;
	case 'f':
		arguments->daemon = false;
		break;
	case 'b':
		arguments->daemon = true;
		break;
	case 'i':
		arguments->ignorefail = true;
		break;
	case 's':
		if (sscanf(arg, "%i", &arguments->random_step) == 0)
			argp_usage(state);
		break;
	case 'W': {
		int n;
		if ((sscanf(arg, "%i", &n) == 0) || (n < 0) || (n > 4096))
			argp_usage(state);
		else
			arguments->fill_watermark = n;
		break;
	}
	case 'q':
		arguments->quiet = true;
		break;
	case 'v':
		printf("%s\n", argp_program_version);
		return -EAGAIN;
	case 'e': {
		int e;
		if ((sscanf(arg,"%i", &e) == 0) || (e < 0) || (e > 8))
			argp_usage(state);
		else
			arguments->entropy_count = e;
		break;
	}
	default:
		return ARGP_ERR_UNKNOWN;
	}

	return 0;
}

static struct argp argp = { options, parse_opt, NULL, doc };


static int update_kernel_random(struct rng *rng, int random_step,
	unsigned char *buf, fips_ctx_t *fipsctx_in)
{
	unsigned char *p;
	int fips;

	fips = fips_run_rng_test(fipsctx_in, buf);
	if (fips)
		return 1;

	for (p = buf; p + random_step <= &buf[FIPS_RNG_BUFFER_SIZE];
		 p += random_step) {
		if (!server_running)
			return 0;
		random_add_entropy(p, random_step);
		random_sleep();
	}
	return 0;
}

static void do_loop(int random_step)
{
	unsigned char buf[FIPS_RNG_BUFFER_SIZE];
	int no_work;
	bool work_done;

	for (no_work = 0; no_work < 100; no_work = (work_done ? 0 : no_work+1)) {
		struct rng *iter;
		int i, retval;

		work_done = false;
		for (i = 0; i < ENT_MAX; ++i)
		{
			int rc;
			/*printf("I is %d\n", i);*/
			iter = &entropy_sources[i];
		retry_same:
			if (!server_running)
				return;

			if (iter->disabled)
				continue;	/* failed, no work */

			message(LOG_DAEMON|LOG_DEBUG, "Reading entropy from %s\n", iter->rng_name);

			retval = iter->xread(buf, sizeof buf, iter);
			if (retval)
				continue;	/* failed, no work */

			work_done = true;

			rc = update_kernel_random(iter, random_step,
					     buf, iter->fipsctx);
			if (rc == 0) {
				iter->success++;
				if (iter->success >= RNG_OK_CREDIT) {
					if (iter->failures)
						iter->failures--;
					iter->success = 0;
				}
				/* succeeded */
				continue;
			}

			iter->failures++;
			if (iter->failures <= MAX_RNG_FAILURES/4) {
				/* FIPS tests have false positives */
				goto retry_same;
			}

			if (iter->failures >= MAX_RNG_FAILURES && !ignorefail) {
				if (!arguments->quiet)
					message(LOG_DAEMON|LOG_ERR,
					"too many FIPS failures, disabling entropy source\n");
				if (iter->close)
					iter->close(iter);
				iter->disabled = true;
			}
		}
	}

	if (!arguments->quiet)
		message(LOG_DAEMON|LOG_ERR,
		"No entropy sources working, exiting rngd\n");
}

static void term_signal(int signo)
{
	server_running = false;
}

static int discard_initial_data(struct rng *ent_src)
{
	/* Trash 32 bits of what is probably stale (non-random)
	 * initial state from the RNG.  For Intel's, 8 bits would
	 * be enough, but since AMD's generates 32 bits at a time...
	 *
	 * The kernel drivers should be doing this at device powerup,
	 * but at least up to 2.4.24, it doesn't. */
	unsigned char tempbuf[4];
	ent_src->xread(tempbuf, sizeof(tempbuf), ent_src);

	/* Return 32 bits of bootstrap data */
	ent_src->xread(tempbuf, sizeof(tempbuf), ent_src);

	return tempbuf[0] | (tempbuf[1] << 8) |
		(tempbuf[2] << 16) | (tempbuf[3] << 24);
}

void close_all_entropy_sources()
{
	int i;
	for (i=0; i < ENT_MAX; i++)
		if (entropy_sources[i].close && entropy_sources[i].disabled == false) {
			entropy_sources[i].close(&entropy_sources[i]);
			free(entropy_sources[i].fipsctx);
	}
}

int main(int argc, char **argv)
{
	int i;
	int ent_sources = 0;
	pid_t pid_fd;

	openlog("rngd", 0, LOG_DAEMON);

	/* Get the default watermark level for this platform */
	arguments->fill_watermark = default_watermark();

	/* Parsing of commandline parameters */
	if (argp_parse(&argp, argc, argv, 0, 0, arguments) < 0)
		return 1;

	if (arguments->list) {
		int found = 0;
		printf("Entropy sources that are available but disabled\n");
		for (i=0; i < ENT_MAX; i++) 
			if (entropy_sources[i].init && entropy_sources[i].disabled == true) {
				found = 1;
				printf("%d: %s\n", i, entropy_sources[i].rng_name);
			}
		if (!found)
			printf("None");
		msg_squash = true;
	} else
		printf("\nInitalizing available sources\n");

	/* Init entropy sources */
	
	for (i=0; i < ENT_MAX; i++) {
		if (entropy_sources[i].init && entropy_sources[i].disabled == false) {
			if (!entropy_sources[i].init(&entropy_sources[i])) {
				ent_sources++;
				entropy_sources[i].fipsctx = malloc(sizeof(fips_ctx_t));
				fips_init(entropy_sources[i].fipsctx, discard_initial_data(&entropy_sources[i]));
			} else {
				if (!arguments->quiet)
					message(LOG_ERR | LOG_DAEMON, "Failed to init entropy source %d: %s\n",
						i, entropy_sources[i].rng_name);
				entropy_sources[i].disabled = true;
			}
		}
	}

	if (arguments->list) {
		int rc = 1;
		msg_squash = false;
		printf("Available and enabled entropy sources:\n");
		for (i=0; i < ENT_MAX; i++) 
			if (entropy_sources[i].init && entropy_sources[i].disabled == false) {
				rc = 1;
				printf("%d: %s\n", i, entropy_sources[i].rng_name);
			}

		close_all_entropy_sources();
		return rc;
	}

	if (!ent_sources) {
		if (!arguments->quiet) {
			message(LOG_DAEMON|LOG_ERR,
				"can't open any entropy source");
			message(LOG_DAEMON|LOG_ERR,
				"Maybe RNG device modules are not loaded\n");
		}
		return 1;
	}
	/* Init entropy sink and open random device */
	init_kernel_rng(arguments->random_name);

	if (arguments->daemon) {
		am_daemon = true;

		if (daemon(0, 0) < 0) {
			if(!arguments->quiet)
				fprintf(stderr, "can't daemonize: %s\n",
				strerror(errno));
			return 1;
		}

		/* require valid, locked PID file to proceed */
		pid_fd = write_pid_file(arguments->pid_file);
		if (pid_fd < 0)
			return 1;

		signal(SIGHUP, SIG_IGN);
		signal(SIGPIPE, SIG_IGN);
		signal(SIGINT, term_signal);
		signal(SIGTERM, term_signal);
	}
	if (arguments->ignorefail)
		ignorefail = true;

	do_loop(arguments->random_step);

	close_all_entropy_sources();

	if (pid_fd >= 0)
		unlink(arguments->pid_file);

	return 0;
}
